java/arcs/android/crdt/CrdtParcelables.kt

```diff
@@ -12,95 +12,42 @@
-    ParcelableCrdtOperation<T>, CrdtOperationAtTime
 
 /** Writes [CrdtData] to the [Parcel]. */
-@Suppress("UNCHECKED_CAST")
-fun Parcel.writeModelData(model: CrdtData?, flags: Int) {
-    when (model) {
-        null -> writeTypedObject(null, flags)
-        is CrdtCount.Data ->
-            writeTypedObject((model as? CrdtCount.Data)?.toParcelable(), flags)
-        is CrdtSet.Data<*> ->
-            writeTypedObject((model as? CrdtSet.Data<Referencable>)?.toParcelable(), flags)
-        is CrdtSingleton.Data<*> ->
-            writeTypedObject((model as? CrdtSingleton.Data<Referencable>)?.toParcelable(), flags)
-        is CrdtEntity.Data ->
-            TODO("Implement me when ParcelableEntity is ready")
-        else -> throw IllegalArgumentException("Unsupported CrdtData type: ${model.javaClass}")
-    }
+fun Parcel.writeModelData(model: CrdtData?) {
+    writeProto(model?.toProto())
 }
 
-/** Reads [CrdtData] from the [Parcel], using the [expectedType] as a hint. */
-fun Parcel.readModelData(expectedType: ParcelableCrdtType): CrdtData? =
-    readTypedObject(
-        requireNotNull(expectedType.crdtDataCreator) {
-            "No crdtDataCreator for type $expectedType"
-        }
-    )?.actual
+/** Reads [CrdtData] from the [Parcel]. */
+fun Parcel.readModelData(): CrdtData? =
+    readProto(CrdtDataProto.getDefaultInstance())?.toData()
 
 /** Writes a [CrdtOperation] to the [Parcel]. */
-@Suppress("UNCHECKED_CAST")
-fun Parcel.writeOperation(operation: CrdtOperation?, flags: Int) {
-    when (operation) {
-        null -> writeTypedObject(null, flags)
-        is CrdtCount.Operation ->
-            writeTypedObject((operation as? CrdtCount.Operation)?.toParcelable(), flags)
-        is CrdtSet.Operation<*> ->
-            writeTypedObject((operation as? CrdtSet.Operation<Referencable>)?.toParcelable(), flags)
-        is CrdtSingleton.Operation<*> ->
-            writeTypedObject(
-                (operation as? CrdtSingleton.Operation<Referencable>)?.toParcelable(), flags)
-        is CrdtEntity.Operation ->
-            TODO("Implement me when ParcelableEntity is ready")
-        else ->
-            throw IllegalArgumentException("Unsupported CrdtOperation type: ${operation.javaClass}")
-    }
+fun Parcel.writeOperation(operation: CrdtOperation?) {
+    writeProto(operation?.toProto())
 }
 
 /** Reads a [CrdtOperation] from the [Parcel]. */
-fun Parcel.readOperation(expectedType: ParcelableCrdtType): CrdtOperation? =
-    readTypedObject(
-        requireNotNull(expectedType.crdtOperationCreator) {
-            "No crdtOperationCreator for type $expectedType"
-        }
-    )?.actual
+fun Parcel.readOperation(): CrdtOperation? =
+    readProto(CrdtOperationProto.getDefaultInstance())?.toOperation()
 
 /** Writes a [List] of [CrdtOperation]s to the [Parcel]. */
-fun Parcel.writeOperations(operations: List<CrdtOperation>, flags: Int) {
+fun Parcel.writeOperations(operations: List<CrdtOperation>) {
```

Comment:
have you considered having multiple operations inside a single CrdtOperationsProto (instead then CrdtOperationProto containing only one op)?

Reply:
It's only used in once place, and I'm hoping to get rid of that usage anyway